File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/react-router-valid-route-object.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticRouteProperty } from "../../utils/get-static-route-property.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isStaticReactRouterRouteObject } from "../../utils/is-static-react-router-route-object.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { wrapReactRouterRule } from "../../utils/wrap-react-router-rule.js";

const EXCLUSIVE_ROUTE_PROPERTY_PAIRS: ReadonlyArray<readonly [string, string]> = [
  ["Component", "element"],
  ["ErrorBoundary", "errorElement"],
  ["HydrateFallback", "hydrateFallbackElement"],
];

export const reactRouterValidRouteObject = wrapReactRouterRule(
  defineRule({
    id: "react-router-valid-route-object",
    title: "Contradictory route object",
    tags: ["test-noise"],
    requires: ["react-router:6.4"],
    severity: "error",
    recommendation:
      "Remove incompatible route properties so each route has one unambiguous rendering contract.",
    create: (context: RuleContext) => ({
      ObjectExpression(node: EsTreeNodeOfType<"ObjectExpression">) {
        if (!isStaticReactRouterRouteObject(context, node)) return;
        const indexProperty = getStaticRouteProperty(node, "index");
        if (
          indexProperty !== null &&
          isNodeOfType(indexProperty.value, "Literal") &&
          indexProperty.value.value === true
        ) {
          const incompatibleProperty =
            getStaticRouteProperty(node, "children") ?? getStaticRouteProperty(node, "path");
          if (incompatibleProperty !== null) {
            context.report({
              node: incompatibleProperty,
              message: "An index route cannot also declare children or a path.",
            });
          }
        }
        for (const [componentPropertyName, elementPropertyName] of EXCLUSIVE_ROUTE_PROPERTY_PAIRS) {
          if (
            getStaticRouteProperty(node, componentPropertyName) === null ||
            getStaticRouteProperty(node, elementPropertyName) === null
          ) {
            continue;
          }
          context.report({
            node,
            message: `Route declares both ${componentPropertyName} and ${elementPropertyName}; only one is used.`,
          });
        }
      },
    }),
  }),
);

File: native/oxlint/rules/react-router-valid-route-object.rs
use oxc_ast::{ast::Expression, AstKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
    AstNode,
};

const EXCLUSIVE_ROUTE_PROPERTY_PAIRS: [(&str, &str); 3] = [
    ("Component", "element"),
    ("ErrorBoundary", "errorElement"),
    ("HydrateFallback", "hydrateFallbackElement"),
];

#[derive(Debug, Default, Clone)]
pub struct ReactRouterValidRouteObject;

declare_oxc_lint!(
    /// Disallow incompatible properties on a React Router route object.
    ReactRouterValidRouteObject,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow contradictory React Router route objects.",
);

impl Rule for ReactRouterValidRouteObject {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_test_noise_file(ctx) && is_react_router_file_active(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::ObjectExpression(route_object) = node.kind() else {
            return;
        };
        if !is_static_react_router_route_object(route_object, ctx) {
            return;
        }
        let is_index_route = get_static_route_property(route_object, "index").is_some_and(|property| {
            matches!(&property.value, Expression::BooleanLiteral(literal) if literal.value)
        });
        if is_index_route {
            if let Some(children_property) =
                get_active_route_property(route_object, "children", ctx)
            {
                ctx.diagnostic(
                    OxcDiagnostic::warn("An index route cannot also declare children.")
                        .with_label(children_property.span),
                );
            }
        }
        for (component_property_name, element_property_name) in EXCLUSIVE_ROUTE_PROPERTY_PAIRS {
            if get_active_route_property(route_object, component_property_name, ctx).is_none()
                || get_active_route_property(route_object, element_property_name, ctx).is_none()
            {
                continue;
            }
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Route declares both {component_property_name} and {element_property_name}; only one is used."
                ))
                .with_label(route_object.span),
            );
        }
    }
}

fn get_active_route_property<'a>(
    route_object: &'a oxc_ast::ast::ObjectExpression<'a>,
    property_name: &str,
    ctx: &LintContext<'a>,
) -> Option<&'a oxc_ast::ast::ObjectProperty<'a>> {
    get_static_route_property(route_object, property_name)
        .filter(|property| !is_definitely_falsy_expression(&property.value, ctx))
}
